transient/sshfs.py

```diff
@@ -115,127 +118,131 @@ def provision_system(timeout: int, ssh_config: ssh.SshConfig, is_slow: bool) ->
-                local_user=local_user,
-                local_password=local_password,
-                ssh_config=ssh_config,
-                is_provisioned=True,
-            )
-        else:
-            raise RuntimeError(
-                f"Unable to provision and SSHFS mount failed with: {stderr}"
+        # Because sshfs monopolizes stdout, the progress marker goes to stderr
+        ssh_command = f"""
+            set -e
+            echo TRANSIENT_SSH_COMPLETE >&2
+            sudo mkdir -p {self.remote_dir}
+            {sshfs_command}
+        """
+
+        sshfs_config = self.ssh_config.override(
+            args=["-T", "-o", "LogLevel=ERROR"] + self.ssh_config.args
+        )
+        client = ssh.SshClient(sshfs_config, command=ssh_command)
+
+        sftp_proc = subprocess.Popen(
+            get_sftp_server(name=sshfs_config.sftp_server_bin_name),
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            preexec_fn=lambda: linux.set_death_signal(signal.SIGTERM),
+        )
+
+        with self.sshfs_sem:
+            logging.info(f"Sending sshfs mount command '{sshfs_command}'")
+            ssh_proc = client.connect(
+                timeout=self.connect_timeout,
+                stdin=sftp_proc.stdout,
+                stdout=sftp_proc.stdin,
+                stderr=subprocess.PIPE,
             )
-    except subprocess.TimeoutExpired:
-        # The timeout expired (as expected), but because we 'set -e', this means
-        # we must be in the state where we're hanging after the logout. So kill
-        # the connection from our end, the sshfs process will continue on the
-        # guest side.
-        conn.terminate()
 
-        # There is a chance the sshfs process hung, so check for the sentinel text
-        raw_stdout, raw_stderr = conn.communicate()
-        stdout = raw_stdout.decode("utf-8")
-        stderr = raw_stderr.decode("utf-8")
-        if "TRANSIENT_SSHFS_DONE" not in stdout:
-            if is_slow:
-                # We timed out without getting to the 'echo' in the script, even with
-                # extra time. Just give up.
-                raise RuntimeError(f"SSHFS mount timed out: {stderr}")
-            else:
-                logging.warning(
-                    "sshfs mount did not complete. Trying with longer timeout"
-                )
-                # Try again with a longer timeout. The system may just be extremely
-                # slow (like qemu with 1 core and no kvm acceleration)
-                do_sshfs_mount(
-                    connect_timeout=connect_timeout,
-                    local_dir=local_dir,
-                    remote_dir=remote_dir,
-                    local_user=local_user,
-                    local_password=local_password,
-                    ssh_config=ssh_config,
-                    is_slow=True,
+            # Everything from here on out simply verifies that nothing went wrong.
+            if ssh_proc.stderr is not None:
```

Comment:
It can't be None, and this bugs me too. However, mypy doesn't look any further than the stated return type of connect(), which superficially says it might have a null stderr. Because mypy can't figure out that stderr will never be None, `make check` fails without this.

I did, however, figure out that changing from an if to an assert allows me to reduce the indentation of the lines in that block without provoking mypy.